File: 44.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Pentagon numbers". The problem
is: Find the pair of pentagonal numbers, Pj and Pk, for which their sum and
difference are pentagonal and D = |Pk - Pj| is minimised; what is the value of
D?
"""

from __future__ import division


def main():
    """Pentagon numbers"""

    is_pentagonal = {1: True}
    pentagons = [1]
    diff = 0
    n = 1
    while not diff:
        n += 1
        next_pentagon = n * (3 * n - 1) // 2
        for pentagon in pentagons:
            candidate = next_pentagon - pentagon
            candidate_diff = abs(pentagon - candidate)
            if candidate in is_pentagonal and candidate_diff in is_pentagonal:
                diff = candidate_diff
        pentagons.insert(0, next_pentagon)
        is_pentagonal[next_pentagon] = True

    print(diff)

if __name__ == '__main__':
    main()

File: 44.js
#!/usr/bin/env node

// Pentagon numbers

var is_pentagonal = {
    1: 1
};
var pentagons = [1];
var diff;
var n = 1;
while (typeof diff == 'undefined') {
    n++;
    var next = n * (3 * n - 1) / 2;
    for (var i = 0; i < pentagons.length; i++) {
        var candidate = next - pentagons[i];
        if (is_pentagonal[candidate] &&
            is_pentagonal[Math.abs(pentagons[i] - candidate)]) {
            diff = Math.abs(pentagons[i] - candidate);
        }
    }
    pentagons.unshift(next);
    is_pentagonal[next] = true;
}

console.log(diff);
